src/nbl/ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -55,6 +56,46 @@ auto CMitsubaMaterialCompilerFrontend::getTexture(const CElementTexture* _elemen
     return { nullptr, nullptr, scale };
 }
 
+
+auto CMitsubaMaterialCompilerFrontend::getEmissionProfile(const CElementEmissionProfile* _element) const -> tex_ass_type
+{
+    asset::IAsset::E_TYPE types[2]{ asset::IAsset::ET_IMAGE_VIEW, asset::IAsset::ET_TERMINATING_ZERO };
+    const auto key = SContext::emissionProfileCacheKey(*_element);
+    auto viewBundle = m_loaderContext->override_->findCachedAsset(key, types, m_loaderContext->inner, 0u);
+    if (!viewBundle.getContents().empty())
+    {
+        auto view = core::smart_refctd_ptr_static_cast<asset::ICPUImageView>(viewBundle.getContents().begin()[0]);
+
+        types[0] = asset::IAsset::ET_SAMPLER;
+        const std::string samplerKey = m_loaderContext->samplerCacheKey(m_loaderContext->emissionProfileSamplerParams(*_element));
+        auto samplerBundle = m_loaderContext->override_->findCachedAsset(samplerKey, types, m_loaderContext->inner, 0u);
+        assert(!samplerBundle.getContents().empty());
+        auto sampler = core::smart_refctd_ptr_static_cast<asset::ICPUSampler>(samplerBundle.getContents().begin()[0]);
+
+        return { view, sampler, 1.0f };
+    }
+    return { nullptr, nullptr, 1.0f };
+}
+
+CMitsubaMaterialCompilerFrontend::EmitterNode* CMitsubaMaterialCompilerFrontend::createEmitterNode(asset::material_compiler::IR* ir, const CElementEmitter* _emitter, core::matrix4SIMD transform) {
+    assert(_emitter->type == CElementEmitter::AREA);
+    auto* res = ir->allocNode<asset::material_compiler::IR::CEmitterNode>();
+    if (_emitter->area.emissionProfile) {
+        asset::material_compiler::IR::CEmitterNode::EmissionProfile profile;
+        asset::material_compiler::IR::INode::STextureSource tex;
+        std::tie(tex.image, tex.sampler, tex.scale) = getEmissionProfile(_emitter->area.emissionProfile);
+        assert(tex.image && "emission profile not loaded");
+        profile.texture = tex;
+        profile.normalizeEnergy = _emitter->area.emissionProfile->normalizeEnergy;
+        auto inverseTransform = core::transpose(core::concatenateBFollowedByA(transform, _emitter->area.emissionProfile->transform.matrix));
+        profile.left = inverseTransform[0];
+        profile.up = inverseTransform[1];
```

Comment:
hmm this is not robust, we need to handle either being LH/RH, non uniform scaling, etc.

Since you can't post-facto anisotropically scale a spherical distribution (maybe you can if you separate out the tensors, but lets not go down a rabbit hole), we require the 3x3 matrix to be orthonormal

I'd extract the Z+ (target axis) first, then build an orthogonal X+ around it and store that, with 1 bit denoting whether Y+ is supposed to be RH or LH relative.

you're missing the orthogonalization step making sure that Z+ and X+ are orthognal and unit length.


hmmm, I'm thinking, maybe we should just store the thing as a quaternion? its compressible into a single RGB21_UNORM because the length encodes the rotation around the axis, and you end up with 1 bit to denote if the Y+ is supposed to be LH or RH w.r.t Z+ and X+.

Reply:
actually quat probably not the best choice since I'll want OBB MIS later, do what you want here, just make sure Z+ and X+ are orho and you can handle Y+ needing flip depending on LH vs RH (which you can infer from the determinant of `invereseTransform`

Reply:
We probably have to orthogonalize matrix as well if transform of emitter object does weird thing. I don't see orthogonalize function in core tho.

Reply:
> We probably have to orthogonalize matrix as well if transform of emitter object does weird thing. I don't see orthogonalize function in core tho.

there isn't one, I'd check that the `inverseTransform` determinant is sufficiently far away from 0, else turn off the emission profile and log a warning/error

then use the scale extraction method of `core::matrix3x4SIMD`, if the scale isn't sufficiently uniform (don't do exact comparisons please cause fp32) for each axis, output a warning and ortho-normalize everything around the `view` (the straight up angle in the IES) and then the `hAngle=0` direction